accepted/PSR-7-http-message-meta.md

```diff
@@ -691,3 +691,14 @@ Further characters or sequences in header values should be rejected according
 to the HTTP specification.
 
 [1]: https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
+
+### 7.2 urlencode User Info
+
+In any release, `UriInterface::withUserInfo` MUST NOT double urlencode the username
+and password if they are already urlencoded.
```

Comment:
I don't see how this can work. Is `h%65llo` already urlencoded or not?

Reply:
What if username *is* `h%65llo` and it should be encoded to `h%2565llo`?

Reply:
an actual username that looks like it is already urlencoded but is actually meant literal will need to be urlencoded by the caller already. 

this already is the case with all of the above mentioned implementations. we could mention that in the errata and maybe add it in the design decisions section.

in a totally strict world, we would have a value object for the username and password that knows if urlencoding already happened or not, but that would be a much bigger change and imho be overcomplicated. are you aware of psr7 implementations that behave differently, where for example passing `h%65llo` would be sent to the server as `h%2565llo` ?

Reply:
> are you aware of psr7 implementations that behave differently?

I am not. I asked this question based on what I would expect as a user. And my first thought regarding your response was “this is bad” and Rob already explained why. Making *assumptions* based on whether some input data is already encoded or not is a common cause of security vulnerabilities. Sanitization / encoding is context dependent and thus must always happen at the output.

As a user I would expect the following:

1. I input some userinfo string in raw form.
2. If the userinfo string is embedded into the URI (i.e. in `__toString()`) it is appropriately encoded.
3. If I extract the userinfo string (e.g. because I want to provide it within an `Authorization` header), then I get the original raw form.

It's unfortunate that the `withUserInfo()` method takes a separate username and password, but the `getUserInfo()` method returns the full string. The RFC 3986 treats the userinfo as an opaque blob (which means that `getUserInfo()` technically behaves correctly).

Reply:
your expection will not be met by any of the above implementations, because all of them urlencode directly in withUserInfo. given that, i think it will be difficult to change the behaviour, even though i agree that getUserInfo returning something else than was passed into withUserInfo sounds unexpected. 

we could add the clarification to getUserInfo that the username and password will be urlencoded, regardless of whether they have been passed to withUserInfo urlencoded or not.

apart from that amendment, i don't see anything else we can do about your input. changing this in all implementations is unrealistic and would make psr7 1.* and 2.* incompatible (on a small thing, but strictly incompatible). if somebody was already doing what you describe, they should right now do urldecode on the getUserInfo - if implementations would change it, the users must not decode anymore, or the `h%65llo` case would break.

Reply:
> your expection will not be met by any of the above implementations, because all of them urlencode directly in withUserInfo.

Yes, I understand that.

> apart from that amendment, i don't see anything else we can do about your input.

My primary intention with the remark was that we *carefully* think about what to do here. I do not have a good answer for now, but I do not believe that this erratum as it is written is the right answer.

> would make psr7 1.* and 2.* incompatible 

Disregard versioning from your mental model. PSRs are immutable once written. The versioning is only about the technical implementation of the interfaces and must only implement what is already written in the PSR (i.e. proper types). See: https://www.php-fig.org/bylaws/psr-evolution/

Reply:
> an actual username that looks like it is already urlencoded but is actually meant literal will need to be urlencoded by the caller already.

So the only reliable way of using this standard is to always encode username before passing it to `withUserInfo()`. In that case, what is the point of this "detect encoding" guidelines, if there is only a one safe way using this standard?

> are you aware of psr7 implementations that behave differently, where for example passing `h%65llo` would be sent to the server as `h%2565llo` ?

I'm not, but from my experience any this kind of conditional encoding solution is can of worms and sooner or later will lead to a problems and frustration. IMO standards like this should be strict - you either always encode, or never. "guess whether string is encoded or not" should not be a part of any standard :).

Reply:
Meanwhile I change the way nyholm/psr7 handles this, see https://github.com/Nyholm/psr7/pull/221, the reason why is discussed in https://github.com/Nyholm/psr7/issues/219, which gives some useful hint: because of the behavior guzzle/laminas have, they're not compatible with e.g. buzz NTLM auth. I argued there that maybe it's buzz that's not using the user info correctly, as it might be required to rawurldecode it before using it. But nothing in the PSR hints about this to this is a but dubious.

The point that triggered the discussion on nyholm/psr7 is the fact that passing a reserved char (in the sense of rfc3986) as user/password produces a broken URI.

If the PSR could be updated to clarify how this is supposed to behave, that'd close the topic.

One way is what other libs do:
- url-encode most chars. But this breaks use cases like buzz'.
- throw an exception when a reserved char is passed. This ship has sailed.
- specify that *reserved chars* MUST be encoded. This is the path taken by nyholm/psr7 as is preserves the use case of buzz.

Reply:
> throw an exception when a reserved char is passed. This ship has sailed.

Even if the ship wouldn't have sailed, that never be correct: The `%` is a reserved character, because it is part of percent-encoding syntax. Thus if reserved characters would be rejected, then neither percent-encoded characters could be passed (because they would contain the reserved `%`), nor a raw `%`.

Given that PSR-7 is clear that `getQuery()` must return percent-encoded values, and that `withQuery()` must accept both encoded and non-encoded values, I tend to agree that this is also correct behavior for the userInfo. For users this means that they effectively always need to manually percent-encode first, because otherwise `h%65llo` would be ambiguous. Is it `hello` with an uselessly encoded `e` or is this `h`, `%`, `65llo`?

The UriInterface really is the “least good” part” of PSR-7, I remember some previous discussion on PHP-FIG discord that I participated in (May, 12 2022) where something else was unclear.

Reply:
> I dont have to encode my query parameters either so why would I do it for the user Info?

@boesing You actually have to:

```
$uri = new Uri('http://127.0.0.1:8080');
$uri = $uri->withQuery('foo=b%61r');
```

Is this a query parameter `foo` with value `bar` or a query parameter foo with value `b%61r`? PSR-7 is clear it's the former, so if you want the latter, you have to encode it. Or ideally use `http_build_query()` which does the right thing.

Reply:
Just to prevent building on a wrong assumption: `%` is *not* a reserved character.

The [ANBF defines them as](https://www.rfc-editor.org/rfc/rfc3986#appendix-A):
```txt
   reserved      = gen-delims / sub-delims
   gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```

Reply:
@nicolas-grekas Sorry, you are right of course. I should have checked this once more, instead of assuming. I've striked out that incorrect part of my comment. The remainder should be correct, though.

Reply:
Yep. So to me the PSR should define the behavior for reserved chars, because they are the ones that are really problematic for interop.

Reply:
> > I dont have to encode my query parameters either so why would I do it for the user Info?
> 
> @boesing You actually have to:
> 
> ```
> $uri = new Uri('http://127.0.0.1:8080');
> $uri = $uri->withQuery('foo=b%61r');
> ```
> 
> Is this a query parameter `foo` with value `bar` or a query parameter `foo` with value `b%61r`? PSR-7 is clear it's the former, so if you want the latter, you have to encode it. Or ideally use `http_build_query()` which does the right thing.

Ha, dunno why I thought it accepts array, maybe good old non-PSR day memory. sorry.

Reply:
> So to me the PSR should define the behavior for reserved chars

Except this also doesn't quite cut it, because the “setter” takes `username` and `password` separately, instead of a userinfo string. The `:` is a gen-delim and thus reserved, but specifically allowed to exist as a raw value within the userinfo. As a percent-encoded reserved character is not equivalent to the unencoded character.

Now RFC 3986 does not define that there the userinfo is a username and password separated with a colon. It defines it as a sequence of opaque bytes. Thus by separately encoding each of the username and password, I am unable to specify a userinfo that contains three colon separated parts. The previous implementation of nyholm/psr7 would have allowed for that (but then misconstructed the resulting URI).

Reply:
Actually the RFC does tell about the user colon password:)
But you're right, the missing spec is about the individual components.

Reply:
> But you're right, the missing spec is about the individual components.

Yes, because the format is scheme-specific (as mentioned at the beginning of RFC 3986#3.2.1).

(This also made me realize a fun aspect of the `mailto:` scheme. The local-part of the email address is actually the `userinfo` part of the URI and that also nicely explains why the userinfo is part of the authority).

---------------

Regarding my confusion with `%` reserved or not:

Even if I was incorrect about the `%` being a reserved character, it nonetheless presents an issue with regard to specifying the behavior, as highlighted by my `getQuery()` example. Users have no other option than encoding the passed values themselves, if they want to ensure they are transmitted exactly how they need them to be transmitted. And at that point I'm wondering if this is solvable with an erratum or if the UriInterface should be fixed with a PSR that supersedes PSR-7.

Reply:
As a user of PSR-7, the only thing that makes sense is for implementations of UriInterface to **always** encode what I pass into this method. It should not guess if it is encoded, it should not throw an exception, and it should not assume that is encoded.

The only logical thing that an implementation can do is assume raw input that MUST be encoded.

Reply:
> the only thing that makes sense is for implementations of UriInterface to always encode what I pass into this method

This would prevent round-tripping of values. See the second paragraph in this comment:

https://github.com/zendframework/zend-diactoros/pull/248#issuecomment-322912952

Reply:
It is highly unfortunate that `getUserInfo` does not decode. I have always found it incredibly annoying (and possibly short sighted) that URI parameters were not decoded on output.

However, I stand by argument because round tripping would become:

```php
$uri = $uri->withUserInfo(rawurldecode($previous->getUserInfo()));
```